pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,984 @@
+	bgpPeerPasswords      map[string]string
+	bgpPeerPasswordsMutex sync.RWMutex
+
+	nodeName     string
+	enabledIPv4  bool
+	enabledIPv6  bool
+	podIPv4CIDR  string
+	podIPv6CIDR  string
+	nodeIPv4Addr string
+
+	egressEnabled bool
+
+	newBGPServerFn func(globalConfig *bgp.GlobalConfig) bgp.Interface
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewBGPPolicyController(ctx context.Context,
+	nodeInformer coreinformers.NodeInformer,
+	serviceInformer coreinformers.ServiceInformer,
+	egressInformer crdinformersv1b1.EgressInformer,
+	bgpPolicyInformer crdinformersv1a1.BGPPolicyInformer,
+	endpointSliceInformer discoveryinformers.EndpointSliceInformer,
+	egressEnabled bool,
+	k8sClient kubernetes.Interface,
+	nodeConfig *config.NodeConfig,
+	networkConfig *config.NetworkConfig) (*Controller, error) {
+	c := &Controller{
+		ctx:                       ctx,
+		nodeInformer:              nodeInformer.Informer(),
+		nodeLister:                nodeInformer.Lister(),
+		nodeListerSynced:          nodeInformer.Informer().HasSynced,
+		serviceInformer:           serviceInformer.Informer(),
+		serviceLister:             serviceInformer.Lister(),
+		serviceListerSynced:       serviceInformer.Informer().HasSynced,
+		bgpPolicyInformer:         bgpPolicyInformer.Informer(),
+		bgpPolicyLister:           bgpPolicyInformer.Lister(),
+		bgpPolicyListerSynced:     bgpPolicyInformer.Informer().HasSynced,
+		endpointSliceInformer:     endpointSliceInformer.Informer(),
+		endpointSliceLister:       endpointSliceInformer.Lister(),
+		endpointSliceListerSynced: endpointSliceInformer.Informer().HasSynced,
+		k8sClient:                 k8sClient,
+		bgpPeerPasswords:          make(map[string]string),
+		nodeName:                  nodeConfig.Name,
+		enabledIPv4:               networkConfig.IPv4Enabled,
+		enabledIPv6:               networkConfig.IPv6Enabled,
+		podIPv4CIDR:               nodeConfig.PodIPv4CIDR.String(),
+		podIPv6CIDR:               nodeConfig.PodIPv6CIDR.String(),
+		nodeIPv4Addr:              nodeConfig.NodeIPv4Addr.IP.String(),
+		egressEnabled:             egressEnabled,
+		newBGPServerFn: func(globalConfig *bgp.GlobalConfig) bgp.Interface {
+			return gobgp.NewGoBGPServer(globalConfig)
+		},
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "bgpPolicyGroup"),
+	}
+	c.bgpPolicyInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addBGPPolicy,
+			UpdateFunc: c.updateBGPPolicy,
+			DeleteFunc: c.deleteBGPPolicy,
+		},
+		resyncPeriod,
+	)
+	c.serviceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addService,
+			UpdateFunc: c.updateService,
+			DeleteFunc: c.deleteService,
+		},
+		resyncPeriod,
+	)
+	c.endpointSliceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addEndpointSlice,
+			UpdateFunc: c.updateEndpointSlice,
+			DeleteFunc: nil,
```

Comment:
why it can ignore delete events? doesn't it affect services of local traffic policy?

Reply:
Thanks for reminding that. I forgot the case that an EndpointSlice delete event could happen even if the corresponding Service is not deleted.